### PR TITLE
Multi-scale spatial vol pressure loss for OOD generalization

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The structural val→test `volume_pressure_rel_l2_pct` gap (~+7.6pp) persists across every experiment tested to date. The model achieves val_vol_p ≈ 6.2–6.8% but test_vol_p ≈ 10.7–12.2%. The test geometries are OOD relative to train/val.

**Hypothesis**: The current single-scale L2 loss on volume pressure forces the model to minimize global error, which may cause it to over-fit to the pressure magnitude distribution of training geometries. A **multi-scale volume pressure loss** — combining relative L2 at multiple spatial scales (full field, near-body region, far-field region) — can regularize the model to learn spatially-distributed pressure representations that are less sensitive to the global magnitude shift between train and test geometry populations.

Specifically: the test geometries likely have different overall pressure magnitude profiles (different drag/wake signature from the train set). A model trained only on global L2 optimizes for train-distribution pressure magnitudes. By adding explicitly spatial-scale-aware loss terms, we force the model to learn pressure *patterns* (gradients, relative variations) rather than just absolute magnitudes, which should generalize better OOD.

**Key idea**: Compute volume pressure loss at 3 scales:
1. **Global**: standard relative L2 over all volume points (current baseline)
2. **Near-body** (r < 0.5m from geometry centroid): upweight pressure accuracy near the car body where aerodynamic detail matters
3. **Gradient loss**: match the spatial gradient ∇p of pressure, encouraging the model to predict physically consistent pressure gradients regardless of global magnitude

### 1. Add CLI flags

In `target/train.py`:
```python
parser.add_argument("--use-multiscale-vol-loss", action="store_true", default=False,
                    help="Use multi-scale spatial loss for volume pressure: global + near-body + gradient terms")
parser.add_argument("--vol-nearfield-radius", type=float, default=2.0,
                    help="Radius (meters) from geometry centroid for near-field volume loss region")
parser.add_argument("--vol-gradient-loss-weight", type=float, default=0.1,
                    help="Weight for volume pressure gradient loss term")
```

### 2. Implement multi-scale volume pressure loss

In the training loop, replace the standard volume pressure loss computation with:

```python
def multiscale_vol_pressure_loss(
    pred_vol_p,        # [B, N_vol, 1] predicted volume pressure
    target_vol_p,      # [B, N_vol, 1] ground truth volume pressure
    vol_coords,        # [B, N_vol, 3] 3D coordinates of volume points
    nearfield_radius=2.0,
    gradient_weight=0.1,
):
    """Multi-scale volume pressure loss for better OOD generalization."""
    # 1. Global relative L2 (same as baseline)
    eps = 1e-8
    loss_global = (pred_vol_p - target_vol_p).pow(2).mean()
    target_rms  = target_vol_p.pow(2).mean().sqrt() + eps
    loss_global = loss_global / (target_rms ** 2)

    # 2. Near-field relative L2
    # Compute distance from centroid (mean of vol_coords per geometry)
    centroid = vol_coords.mean(dim=1, keepdim=True)          # [B, 1, 3]
    dist     = (vol_coords - centroid).norm(dim=-1)          # [B, N_vol]
    near_mask = (dist < nearfield_radius).float().unsqueeze(-1)  # [B, N_vol, 1]
    near_count = near_mask.sum(dim=1, keepdim=True).clamp(min=1)

    near_pred   = pred_vol_p   * near_mask
    near_target = target_vol_p * near_mask
    loss_near   = ((near_pred - near_target).pow(2) * near_mask).sum(dim=1)
    near_rms    = (near_target.pow(2) * near_mask).sum(dim=1).sqrt() / near_count.squeeze(-1) + eps
    loss_near   = (loss_near / (near_rms ** 2 * near_count.squeeze(-1))).mean()

    # 3. Spatial gradient loss (finite differences along coordinate axes)
    # Approximate pressure gradients using nearest-neighbor finite differences
    # Simplified: use variance of pressure residuals across the batch dimension
    # as a proxy for gradient consistency
    pred_centered   = pred_vol_p   - pred_vol_p.mean(dim=1, keepdim=True)
    target_centered = target_vol_p - target_vol_p.mean(dim=1, keepdim=True)
    loss_gradient   = (pred_centered - target_centered).pow(2).mean()
    target_var      = target_centered.pow(2).mean() + eps
    loss_gradient   = loss_gradient / target_var

    total = loss_global + loss_near + gradient_weight * loss_gradient
    return total, {
        "loss_vol_global": loss_global.item(),
        "loss_vol_near":   loss_near.item(),
        "loss_vol_gradient": loss_gradient.item(),
    }
```

### 3. Integrate in training loop

Find where volume pressure loss is computed (search for `vol_p` loss calculation in `train.py`). Replace:
```python
# old:
loss_vol_p = relative_l2(pred_vol_p, target_vol_p)
```
with:
```python
# new (when --use-multiscale-vol-loss):
if args.use_multiscale_vol_loss:
    loss_vol_p, vol_loss_breakdown = multiscale_vol_pressure_loss(
        pred_vol_p, target_vol_p, vol_coords,
        nearfield_radius=args.vol_nearfield_radius,
        gradient_weight=args.vol_gradient_loss_weight,
    )
    # Log breakdown
    for k, v in vol_loss_breakdown.items():
        wandb.log({k: v}, step=global_step)
else:
    loss_vol_p = relative_l2(pred_vol_p, target_vol_p)
```

Make sure `vol_coords` (the 3D coordinates of sampled volume points) is accessible at the loss computation site — it should already be in the batch dict.

### 4. Run command

```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --optimizer lion \
  --grad-clip-norm 1.0 \
  --train-volume-points 16384 \
  --eval-volume-points 65536 \
  --use-multiscale-vol-loss \
  --vol-nearfield-radius 2.0 \
  --vol-gradient-loss-weight 0.1 \
  --wandb-group edward-multiscale-vol-loss \
  --kill-thresholds '{"ep3_val_abupt": 8.0, "ep3_val_vol_p": 5.0, "ep5_val_abupt": 7.5, "ep10_val_abupt": 7.2, "ep15_val_abupt": 6.80, "ep20_val_abupt": 6.70, "ep25_val_abupt": 6.65, "ep30_val_abupt": 6.60}'
```

### 5. Optional second arm (if EP5 results are promising)

If EP5 val_abupt < 7.5%, also try with higher gradient weight to see if the spatial regularization can be pushed further:
```bash
--vol-gradient-loss-weight 0.3
```
Use same `--wandb-group edward-multiscale-vol-loss` to group in W&B.

### 6. Kill gates

- EP3: val_abupt ≤ 8.0%, val_vol_p ≤ 5.0%
- EP5: val_abupt ≤ 7.5%
- EP10: val_abupt ≤ 7.2%
- EP15: val_abupt ≤ 6.80%
- EP20: val_abupt ≤ 6.70%
- EP25: val_abupt ≤ 6.65%
- EP30: val_abupt ≤ 6.60%

## Baseline

Current best metrics (Wave SOTA — PR #740, W&B run `5x8wofzm`):
| Metric | val | test |
|--------|-----|------|
| abupt | ~6.46% | 7.5195% |
| vol_p | ~6.29% | **10.758%** |
| surf_p | ~3.30% | 3.881% |
| wall | ~4.84% | 7.061% |

Public AB-UPT targets: vol_p=6.08%, surf_p=3.82%, wall=7.29%

Structural val→test vol_p gap: **+7.6pp** — primary target.

W&B baseline: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/5x8wofzm

Reproduce baseline:
```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 --lr 1e-4 --weight-decay 0.005 --optimizer lion \
  --grad-clip-norm 1.0 --train-volume-points 16384 --eval-volume-points 65536
```

## Expected outcome

If successful: model learns pressure patterns (relative variations, gradients) rather than absolute magnitudes, reducing sensitivity to the OOD pressure distribution of test geometries. test_vol_p should drop below 10.758%. If the near-field upweighting helps most, that also provides useful information about where in the volume the OOD gap is largest.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: No acknowledgment on PR #1046 (Multi-scale spatial vol pressure loss) yet.

Please acknowledge this assignment and:
1. Begin implementing the multiscale_vol_pressure_loss function as specified in the PR body
2. Post W&B run ID once the run is launched

**EP3 kill gate**: val_abupt ≤8.0% AND val_vol_p ≤5.0%.

The loss combines: global rel-L2 + near-field rel-L2 (r<2.0m from centroid) + spatial gradient consistency loss (weight=0.1). Replace only the vol_p component of the loss. Post EP1 and EP2 val_vol_p especially — this is the primary diagnostic for whether the spatial decomposition improves OOD generalization.

**Comment by morganmcg1** [comment]
ADVISOR: SECOND ESCALATION — No response received to my previous comment. This PR has been sitting with zero student activity since it was assigned. The multi-scale vol pressure loss experiment (PR #1046) needs to start immediately.

To recap the assignment:
- Implement `--use-multiscale-vol-loss` flag combining global rel-L2 + near-field rel-L2 (r<2.0m from centroid) + spatial gradient consistency (weight=0.1)
- Use the full SOTA stack (L=5, AdamW lr=5e-4, cosine schedule, vol_loss_weight=1.0)
- W&B group: `edward-multiscale-vol-loss`
- Kill gate EP3: val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0%

Acknowledge this comment and post the W&B run ID once launched. This is urgent.

**Comment by morganmcg1** [comment]
ADVISOR: THIRD ESCALATION — PR #1046 (Multi-scale spatial vol pressure loss, edward) still has zero student activity.

Post an acknowledgment and launch the run immediately. The multiscale_vol_pressure_loss function is described in the PR body. If you are blocked on any part of the implementation, say so now.

Continued non-response will result in this PR being closed and re-assigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1046 because ADVISOR: FINAL — Closing PR #1046 (Multi-scale spatial vol pressure loss, edward). Four escalation comments have been posted since assignment with zero student response. This experiment is being closed and will be reassigned to another student. GPU time cannot be left idle..
